### PR TITLE
Add `wp atmosphere backfill` WP-CLI command

### LINKED ISSUE #88: Feature Request: WP-CLI command for backfill
### What

Add a `wp atmosphere backfill` WP-CLI command that exposes the same logic as the admin-AJAX backfill in `includes/class-backfill.php`.

Suggested flags:

- `--post-type=<type>` — limit to a single supported post type.
- `--ids=<csv>` — explicit list of post IDs, bypasses the unsynced query.
- `--limit=<n>` — cap on posts processed (default: no cap, unlike the UI's default of 10).
- `--batch=<n>` — batch size for progress reporting.
- `--dry-run` — list posts that would be published without calling `Publisher::publish_post()`.
- `--force` — republish even when the `Document::META_URI` post meta is already set.

### Why

The current backfill is admin-AJAX only. That works fine for a handful of posts clicked through the settings page, but it leaves a few gaps:

- No way to run a backfill headless from a deploy hook, cron, or CI.
- Large catalogs (thousands of posts) are awkward in the browser — page timeouts, refresh required, easy to lose progress.
- No good way to republish a single post on demand while debugging without round-tripping through the UI.
- Multisite or "many installs" rollouts are scriptable from WP-CLI but not from the AJAX endpoint.

In short: anything operational beyond "I have ten old posts I want on Bluesky" benefits from a CLI.

### How

Register the command behind `class_exists( 'WP_CLI' )` so the autoloader stays clean for the web request path. The bulk of the work is wrapping the existing query in `Backfill::handle_count()` and the `Publisher::publish_post()` loop in `Backfill::handle_batch()` — no new business logic. `WP_CLI\Utils\make_progress_bar()` for progress; `WP_CLI::success` / `WP_CLI::warning` per result. Worth pulling the shared query out of `handle_count()` into a static helper so the CLI and AJAX paths can't drift.

### PR DESCRIPTION
**Pull request by kraftbj**
Fixes #88

## Proposed changes:

* Add `wp atmosphere backfill` WP-CLI command (new `includes/class-cli.php`, registered from `atmosphere.php` only when `WP_CLI` is defined). Mirrors the admin-AJAX backfill so backfills can run headless from cron, deploy hooks, or ad-hoc scripts.
* Extract the unsynced-posts query out of `Backfill::handle_count()` into a shared static helper `Backfill::get_unsynced_post_ids( int $limit, array $post_types ): array`. Both the CLI command and the AJAX endpoint call it, so the two surfaces cannot drift on which posts qualify for sync.
* Page the unsynced-posts walk in chunks (default 500, overridable via the new `atmosphere_backfill_query_chunk_size` filter) so large catalogues do not OOM before the limit is checked. Per-chunk `wp_cache_delete_multiple()` evicts the chunk's primed post_meta rows before the next page loads.
* Re-sync flag routes through `Publisher::update_post()` when the post already has a document URI, so existing TIDs and external engagement (likes, reposts, replies) are preserved.
* Strict `--ids` parsing: rejects partially numeric tokens (`1.5`, `123abc`, `1-2`) instead of silently truncating them, and surfaces the rejected list as a fatal error before any publish runs.
* Per-batch object-cache eviction across the entire `--batch` window, not just the boundary post. Bounds memory on long runs.
* Add PHPUnit coverage for the helper and the CLI's input parsing (`tests/phpunit/tests/class-test-backfill.php`).

### Flag set

| Flag | Default | Behavior |
|---|---|---|
| `--post-type=<type>` | all supported | Limit to a single supported post type; bail with `WP_CLI::error` if not in `get_supported_post_types()`. Ignored under `--ids` (warns; validation is also skipped). |
| `--ids=<csv>` | (none) | Explicit list of post IDs; bypasses the unsynced query. Each ID still runs through `is_post_publishable()`; skipped IDs are reported. Partially numeric tokens (`1.5`, `123abc`) abort the run before any publish. |
| `--limit=<n>` | `0` (no cap) | Cap on posts processed. Non-numeric and negative values are rejected. The AJAX UI defaults to 10; the CLI default is unlimited so a scripted run can drain a queue in one invocation. |
| `--batch=<n>` | `25` | Progress-reporting cadence AND per-batch object-cache eviction. Does not change the publish loop semantics. |
| `--dry-run` | off | List posts that would be published; does NOT call `Publisher::publish_post()`. |
| `--force` | off | Re-sync posts that already carry the document URI meta. Posts with a prior successful publish go through `Publisher::update_post()` (in-place `applyWrites#update`, preserves TIDs/URIs/external engagement). Half-state posts (TID set, URI missing) fall back to publish via the same path. |
| `--original-time` | off | Reserved for an upcoming change that adds historical timestamp support (see note below). |

Exit code is non-zero when any publish fails so cron and CI wrappers can detect partial failures from `$?`.

### Note on `--original-time`

The flag is wired but currently a no-op — both branches publish identically. Issue 89 will add `TID::generate_for_time()`, and a follow-up wire-up will make this flag select the historical TID path. Including the flag in this PR keeps the CLI's external contract stable before that work merges, so deploy scripts written today against the documented flag set will not need to change.

### Other information:

- [x] Have you written new tests for your changes, if applicable?

Tests cover the extracted `Backfill::get_unsynced_post_ids()` helper (eligibility rules, post-type filter, limit semantics, paged walk across chunk boundaries, exact-multiple paging, limit-only-counts-unsynced across boundaries, int-only return contract) and `CLI::parse_ids()` (ten provider rows for the documented input contract plus a dedicated `test_parse_ids_rejects_partially_numeric_tokens` covering the codex finding). The CLI command's outer flow is exercised indirectly via the helper; testing `\WP_CLI::*` output through `WP_UnitTestCase` requires stubbing the WP-CLI runtime and was deferred per the project guidance (the test SKILL.md notes that stubbing WP-CLI is only worth it when it stays under ~50 lines). The `Publisher::update_post()` contract that the re-sync path relies on is already covered by `Test_Publisher::test_update_sends_update_writes` and its half-state-fallback siblings.

### Reviewer notes

- `composer lint` is clean on the diff.
- `npm run env-test` was **not** run from this worktree across any review round because this branch's wp-env environment was not the one currently up on the machine (a different worktree's environment was running) and project conventions forbid restarting Docker containers without explicit permission. Please run `npm run env-test -- --filter=Backfill` and `--filter=Publisher` before merging to confirm the new test class passes and the existing Publisher coverage still passes against the re-sync routing change.
- The CLI command was not exercised end-to-end against a live wp-env for the same reason. Manual smoke test instructions are below.

### Review rounds

Three rounds of persona review (correctness, maintainability, project-standards, testing, security, performance) plus one codex pass. Each round's findings landed as 1-3 follow-up commits on top of the original implementation.

**Round 1** — progress-bar guard fix, non-zero exit on errors, paged unsynced query to bound memory, reject negative `--limit`, error on all-invalid `--ids`, warn when `--ids` and `--post-type` are both supplied, split dry-run suffix into separate `_n()` variants, hoist progress tick, per-batch cache eviction, soften `--original-time` help text, expand test coverage (future-status, int return contract, tighter limit-cap test, paged-walk regression, eight `parse_ids` provider rows).

**Round 2** — fix the cache eviction to actually evict the whole batch (not just the boundary post), per-chunk post_meta eviction in `Backfill::get_unsynced_post_ids()`, tighten `--limit` validation against `is_numeric`, replace silent-revert-to-default with `\max( 1, $chunk_size )` clamp, add three boundary-condition tests (exact-multiple paging, order-after-dedup, limit-only-counts-unsynced across a page boundary).

**Round 3 (codex)** — three findings, all applied:
- **Fixed:** Route re-sync flag through `Publisher::update_post()` when META_URI is set (previously called `publish_post()` which issues `applyWrites#create` ops the PDS rejects for existing records).
- **Fixed:** Strict `parse_ids()` rejects partially numeric tokens (`1.5`, `123abc`, `1-2`, `+5`) instead of silently truncating via `(int)`. Returns `{ids, rejected}`; CLI errors with the rejected token list before any publish runs.
- **Fixed:** Skip `--post-type` validation when `--ids` is supplied so an unsupported post-type slug under explicit IDs doesn't block the run (the help text says `--post-type` is ignored there; validation now matches).

Also rolled into the codex round: dropped the redundant `(int) $post_id` cast inside the publish loop now that `parse_ids()` and `Backfill::get_unsynced_post_ids()` both guarantee int[] returns.

### Follow-ups (not blocking)

Out of scope for this PR; noted so the reviewer has visibility:

- Extract a shared `Backfill::publish_post_ids( array $ids, array $opts )` between the CLI and AJAX paths. The AJAX handler's response shape would need to change too; doing it carefully is a separate focused PR.
- PDS rate limiting (`--sleep`, 429 backoff). Real concern, but scope creep for a "make the AJAX backfill scriptable" PR.
- `\WP_CLI::confirm()` when the candidate set is huge.
- LOW polish items the round 3 persona review flagged: small duplicated 3-line patterns inside the publish loop's skip / not-found / not-eligible branches, fixture-helper opportunities in `class-test-backfill.php`, and `--limit=5.5` still truncates silently (the round 3 codex fix only touched `--ids`; a follow-up could apply the same `ctype_digit` discipline to `--limit`).

## Testing instructions:

1. Start wp-env: `npm run env-start`.
2. Run the new PHPUnit class: `npm run env-test -- --filter=Backfill`. All tests should pass.
3. Run the Publisher tests too, to confirm the re-sync routing did not regress anything: `npm run env-test -- --filter=Publisher`.
4. Run the full suite: `npm run env-test`.
5. Connect the plugin to an AT Protocol account.
6. Create several published posts via `wp post create` or the admin UI, leaving them unsynced.
7. Smoke test the CLI:
   ```
   npm run env -- run cli wp atmosphere backfill --dry-run
   npm run env -- run cli wp atmosphere backfill --limit=2
   npm run env -- run cli wp atmosphere backfill --ids=POST_ID                # republishes a synced post via update path
   npm run env -- run cli wp atmosphere backfill --ids=1.5                    # should error before publishing
   npm run env -- run cli wp atmosphere backfill --ids=POST_ID --post-type=does-not-exist  # should warn, NOT error
   npm run env -- run cli wp atmosphere backfill --post-type=does-not-exist   # should error with the supported list
   npm run env -- run cli wp atmosphere backfill --limit=-1                   # should error
   npm run env -- run cli wp atmosphere backfill --limit=abc                  # should error
   npm run env -- run cli wp atmosphere backfill --ids=abc                    # should error (no valid IDs)
   ```
8. For a post that has already been published once: re-run the CLI with the re-sync flag (`--force`), check the Bluesky / standard.site records' URIs, and confirm the AT-URIs are unchanged (proves the update path was taken).
9. Verify the final summary line reports `Synced X of Y posts (Z skipped, W errors).`
10. Verify the process exit code is non-zero when at least one publish errors (`echo $?` after the command).
11. Confirm the admin-AJAX backfill still works as before — the settings page batch count should match `wp atmosphere backfill --dry-run --limit=10`.

### Changelog entry

- [x] Automatically create a changelog entry from the details below.

Already added at `.github/changelog/wpcli-backfill`.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

- [x] Minor

#### Type

- [x] Added - for new features

#### Message

A new WP-CLI command can backfill existing posts to AT Protocol from the command line.

</details>

### CODE REVIEW

File: tests/phpunit/class-wp-cli-command.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * itself is not loaded by the WP test bootstrap; without this stub,
+ * autoloading `Atmosphere\Cli\Backfill_Command` fatals on the missing
+ * parent class. Tests that exercise the command's outer flow are not
+ * in scope — only its static input parser is — so an empty stub is
+ * sufficient.
+ *
+ * @package Atmosphere
+ */
+
+\defined( 'ABSPATH' ) || exit;
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/class-wp-cli-command.php]: Good catch. Fixed in d153157: dropped the `defined( 'ABSPATH' ) || exit;` from the test stub. The stub is required *before* WordPress core's bootstrap defines ABSPATH, so the guard would have exited PHPUnit before any test ran. Tests have been passing because the wp-env / WP-CLI harness defines ABSPATH at a higher layer, but relying on that was fragile — the guard had no place in a test-only shim.

File: includes/cli/class-backfill-command.php [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+		 * detect failure via the process exit code — `success()` would
+		 * exit 0 even when every publish errored, hiding the failure from
+		 * cron and CI wrappers.
+		 */
+		if ( $errors > 0 ) {
+			\WP_CLI::error( $summary );
+		}
+
+		\WP_CLI::success( $summary );
+	}
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/cli/class-backfill-command.php:508]: Fixed in d153157: dry-run summary now reads "Would publish %d of %d posts (%d skipped). Dry run, nothing was sent to AT Protocol." and exits 0 unconditionally (no error-capable publish loop in dry-run).

File: tests/phpunit/tests/class-test-backfill.php [hunk 3]
```diff
@@ -0,0 +1,339 @@
+<?php
+/**
+ * Tests for the Backfill helper and the WP-CLI command's input parser.
+ *
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/tests/class-test-backfill.php:7]: Fixed in d153157: file docblock now points to `tests/phpunit/tests/cli/class-test-backfill-command.php` for the input-parser tests.

File: .agents/skills/code-style/SKILL.md [hunk 4]
```diff
@@ -59,7 +59,7 @@ Always reserve the rkey via meta in `get_rkey()` — that meta key is the marker
 
 **Content / composition:** `atmosphere_content_parser`, `atmosphere_document_content`, `atmosphere_long_form_composition`, `atmosphere_teaser_thread_posts`.
 
-**Gating:** `atmosphere_syncable_post_types`, `atmosphere_should_publish_comment`, `atmosphere_should_sync_reply`, `atmosphere_backfill_limit`, `atmosphere_oauth_redirect_uri`, `atmosphere_client_metadata`.
+**Gating:** `atmosphere_syncable_post_types`, `atmosphere_should_publish_comment`, `atmosphere_should_sync_reply`, `atmosphere_oauth_redirect_uri`, `atmosphere_client_metadata`.
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply .agents/skills/code-style/SKILL.md]: Fixed in d153157: `atmosphere_backfill_query_chunk_size` added to the Gating list.

File: includes/class-backfill.php [hunk 5]
```diff
@@ -15,137 +17,134 @@
...
+	 * `atmosphere_backfill_query_chunk_size` filter, default 500) so a
+	 * 50k-post catalogue does not load every ID and prime every meta
+	 * row before the limit is even checked. Each chunk is primed with
+	 * `update_meta_cache()` so the per-post `get_post_meta()` reads do
+	 * not fire one SELECT per ID. `no_found_rows` skips the
+	 * `SQL_CALC_FOUND_ROWS` overhead — we never need the total count.
+	 *
+	 * @since unreleased
+	 *
+	 * @param int      $limit      Maximum IDs to return. 0 or negative for no cap.
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-backfill.php:60]: `@since unreleased` is the project convention for new public surface — `docs/php-coding-standards.md` documents it and the release script (`bin/release.js`) rewrites the placeholder to the actual version at release time. Leaving the placeholder in place is intentional; replacing it with a concrete version pre-release would be wrong.

File: includes/class-backfill.php [hunk 6]
```diff
@@ -15,137 +17,134 @@
...
+		 * it to trade memory for fewer queries on healthy hosts. The
+		 * filter also gives the test suite a knob for driving the paged
+		 * loop with small fixtures.
 		 *
-		 * Only the most recent unsynced posts (up to this limit) will
-		 * be included in each backfill run. Use 0 or -1 to backfill all.
+		 * @since unreleased
 		 *
-		 * @param int $limit Maximum number of posts. Default 10.
+		 * @param int $chunk_size Posts per page. Default 500.
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/class-backfill.php:84]: Same as the comment above on line 60 — `unreleased` is the documented placeholder; the release script rewrites it at release time.

File: docs/php-coding-standards.md [hunk 7]
```diff
@@ -220,7 +220,6 @@ use function Atmosphere\is_connected;
 \apply_filters( 'atmosphere_syncable_post_types',     array( 'post' ) );
 \apply_filters( 'atmosphere_should_publish_comment',  $bool, $comment );
 \apply_filters( 'atmosphere_should_sync_reply',       $bool, $notification, $post_id );
-\apply_filters( 'atmosphere_backfill_limit',          50 );
 \apply_filters( 'atmosphere_oauth_redirect_uri',      $uri );
 \apply_filters( 'atmosphere_client_metadata',         $metadata );
 ```
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply docs/php-coding-standards.md:226]: Fixed in d153157: `atmosphere_backfill_query_chunk_size` added to the gating-filters list.

**Comment by kraftbj** [comment]
Let me clean this branch up a bit.

File: includes/cli/class-backfill-command.php [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+			 * Strict digit check. `(int) '1.5'` would silently become 1
+			 * and republish a different post than the operator typed —
+			 * exactly the safety bug codex flagged. `ctype_digit` only
+			 * accepts strings of digits 0-9, so leading signs, decimals,
+			 * range syntax (`1-2`), and trailing junk all land in
+			 * `$rejected` and the caller errors out before any publish.
+			 */
+			if ( ! \ctype_digit( $trimmed ) ) {
+				$rejected[] = $part;
+				continue;
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/cli/class-backfill-command.php:558]: Addressed in 090a01a — rejected tokens are now recorded after trimming, so leading/trailing whitespace no longer appears in the error output.

File: includes/class-backfill.php [hunk 9]
```diff
@@ -15,137 +17,134 @@
...
-			if ( ! $post || ! is_post_publishable( $post ) ) {
-				$results[] = array(
-					'id'      => $post_id,
-					'success' => false,
-					'error'   => \__( 'Post not eligible for sync.', 'atmosphere' ),
-				);
-				continue;
+				if ( $limit > 0 && \count( $unsynced ) >= $limit ) {
+					break 2;
+				}
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-backfill.php:153]: Fixed in 6ba1616 — the chunk's primed post_meta is now evicted before the --limit break, so the capped path bounds memory like the full walk.

File: includes/cli/class-backfill-command.php [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+						\__( 'Invalid --limit value "%s": expected 0 (no cap) or a positive integer.', 'atmosphere' ),
+						\is_scalar( $raw_limit ) ? (string) $raw_limit : \gettype( $raw_limit )
+					)
+				);
+			}
+
+			$limit = (int) $raw_limit;
+		} else {
+			$limit = 0;
+		}
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/cli/class-backfill-command.php:155]: Fixed in 9ba340b — --limit now uses a strict ctype_digit check, so 5.5, 1e3, and negatives are rejected instead of being silently truncated.

File: includes/class-backfill.php [hunk 11]
```diff
@@ -15,137 +17,134 @@
...
+					'has_password'           => false,
+					'posts_per_page'         => $chunk_size,
+					'paged'                  => $page,
+					'orderby'                => 'date',
+					'order'                  => 'DESC',
+					'fields'                 => 'ids',
+					'no_found_rows'          => true,
+					'update_post_term_cache' => false,
+				)
+			);
```
[the remark this answers is not in the document]
Comment (kraftbj) [reply includes/class-backfill.php:128]: Fixed in 6ba1616 — the query now orders by date with an ID tiebreaker, so equal-date rows cannot reorder across pages and duplicate or skip an ID.

**Comment by kraftbj** [comment]
Hi Matthias! I'm done touching this branch, so it's all yours again whenever you want to pick it back up.

A couple of heads-ups on the current state:

**History was rewritten (force-pushed).** After our pushes crossed, the old duplicate-merge tangle had come back. I rewrote the branch into a clean linear history with your `090a01a` (i18n + test cleanup) preserved on top. Because the history changed, please re-point your local branch at origin before continuing (fetch, then hard-reset your local `add/wpcli-backfill-ARC-88` onto `origin/add/wpcli-backfill-ARC-88`) so the old commits don't come back.

**Review fixes applied** on top of your commit:

- Deterministic pagination: the unsynced walk now orders by date with an ID tiebreaker, so equal-date rows can't duplicate or skip across pages.
- Stricter flag validation: a valueless `--ids` and a non-integer `--limit` (5.5, 1e3, -1) are now rejected instead of silently targeting post 1 or truncating.
- `--force` on a half-synced post (document URI present but no Bluesky record) is reported as a skip instead of a misleading "Updated" success.
- Fail fast when not connected, and abort the rest of the run on a mid-run connection loss instead of erroring on every remaining post.
- Dry runs now bound memory the same way real runs do, and `--original-time` warns that it's a no-op for now (flag left in place for your follow-up).
- Added integration coverage for the command's `__invoke()` flow. Full suite is green (547 tests) and PHPCS is clean.

I also resolved the open Copilot threads these commits address. Thanks!

File: tests/phpunit/class-wp-cli.php [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+			}
+
+			self::$calls[] = array(
+				'level'   => 'error',
+				'message' => $text,
+			);
+
+			if ( $halt ) {
+				throw new WP_CLI_Halt( \esc_html( $text ) );
+			}
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/class-wp-cli.php:141]: Good catch — the stub now throws the raw `$text` instead of `esc_html( $text )` (with a `phpcs:ignore` for the escape sniff), so a caught `WP_CLI_Halt` message matches the real WP-CLI runtime and is not HTML-entity-encoded. Fixed in 54dddc6.

File: tests/phpunit/class-wp-cli-command.php [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Stub for WP-CLI's command base class.
+ *
+ * Loaded from {@see tests/phpunit/bootstrap.php} so the CLI command
+ * classes under `includes/cli/` can be autoloaded under PHPUnit. WP-CLI
+ * itself is not loaded by the WP test bootstrap; without this stub,
+ * autoloading `Atmosphere\Cli\Backfill_Command` fatals on the missing
+ * parent class. Tests that exercise the command's outer flow are not
+ * in scope — only its static input parser is — so an empty stub is
+ * sufficient.
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/class-wp-cli-command.php]: Updated — the docblock no longer claims outer-flow tests are out of scope. It now notes this stub only supplies the (empty) parent class so the command autoloads, and that the outer flow is exercised via the `class-wp-cli.php` facade stub. Fixed in 54dddc6.

File: tests/phpunit/tests/class-test-backfill.php [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+
+		$this->assertSame( array(), Backfill::get_unsynced_post_ids( 0, array() ) );
+		$this->assertSame( array(), Backfill::get_unsynced_post_ids( 10, array() ) );
+	}
+
+	/**
+	 * Page-boundary exact-multiple regression: when the candidate set
+	 * exactly fills an integer number of chunks, the do/while must
+	 * exit via the `0 === $chunk_count` branch rather than producing a
+	 * spurious extra query (or, worse, missing the last full chunk).
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply tests/phpunit/tests/class-test-backfill.php]: Reworded — the docblock now says the walk does one final empty-page query to confirm completion, then exits via the `0 === $chunk_count` branch, so the last full chunk is neither missed nor returned twice. Fixed in 54dddc6.

File: includes/wp-admin/class-settings-fields.php [hunk 15]
```diff
@@ -101,14 +101,6 @@ public static function register_settings_fields(): void {
 			'atmosphere_publishing'
 		);
```
[the remark this answers is not in the document]
Comment (pfefferle) [reply includes/wp-admin/class-settings-fields.php:102]: Leaving the admin backfill UI fully removed by design — `wp atmosphere backfill` is the documented replacement (the command's `## OPTIONS` / `## EXAMPLES` help plus the changelog entry). We'd rather not reintroduce a settings field/description for it.

**Review by pfefferle — APPROVED** [review]
